ao-contracts/curve-bonded-token/src/handlers/bonding_curve.lua

```diff
@@ -0,0 +1,591 @@
+
+    if bint.__le(S_exp, 0) then
+        msg.reply({
+            Action = 'Buy-Tokens-Error',
+            Error = 'Bonding curve error: S_EXP too low ' .. S_exp
+        })
+
+        return
+    end
+
+    -- Cost = G * [ (s2)^(n+1) - (s1)^(n+1) ] / S^(n+1)
+    local s1_exp = bint.ipow(bint(s1), bint(EXP_N_PLUS1))
+    local s2_exp = bint.ipow(bint(s2), bint(EXP_N_PLUS1))
+
+    local numerator = utils.multiply(FundingGoal, utils.subtract(s2_exp, s1_exp))
+    local cost = utils.divide((numerator), S_exp)
+    LogActivity(msg.Tags['X-Action'], json.encode({ Cost = tostring(cost), AmountSent = tostring(quantityReservesSent) }),
+        "Calculated cost of buying tokens for Reserves sent")
+    if bint.__lt(bint(quantityReservesSent), bint(math.ceil(utils.toNumber(cost)))) then
+        LogActivity(msg.Tags['X-Action'],
+            json.encode({ Cost = tostring(cost), AmountSent = tostring(quantityReservesSent) }),
+            "Insufficient funds sent to buy")
+        local refundSuccess = RefundHandler(quantityReservesSent, sender, reservePID)
+
+        if not refundSuccess then
+            LogActivity(msg.Tags['X-Action'],
+                json.encode({ Cost = tostring(cost), AmountSent = tostring(quantityReservesSent) }),
+                "Refund failed")
+            return
+        end
+
+        ao.send({
+            Target = sender,
+            Action = "Refund-Notice",
+            Quantity = tostring(quantityReservesSent),
+        })
+
+        msg.reply({
+            Cost = tostring(cost),
+            AmountSent = tostring(quantityReservesSent),
+            Action = 'Buy-Tokens-Error',
+            Error = 'Insufficient funds sent to buy'
+        })
+
+        return
+    end
+
+    local mintResp = ao.send({ Target = RepoToken.processId, Action = "Mint", Tags = { Quantity = utils.toBalanceValue(tokensToBuyInSubUnits), Recipient = sender } })
+        .receive()
+
+    if mintResp.Tags['Action'] ~= 'Mint-Response' then
+        LogActivity(msg.Tags['X-Action'],
+            json.encode({ Cost = tostring(cost), AmountSent = tostring(quantityReservesSent) }),
+            "Failed to mint tokens")
+        local refundSuccess = RefundHandler(quantityReservesSent, sender, reservePID)
+
+        if not refundSuccess then
+            LogActivity(msg.Tags['X-Action'],
+                json.encode({ Cost = tostring(cost), AmountSent = tostring(quantityReservesSent) }),
+                "Refund failed after failed mint")
+            return
+        end
+
+        ao.send({
+            Target = sender,
+            Action = "Refund-Notice",
+            Quantity = tostring(quantityReservesSent),
+        })
+        msg.reply({
+            Action = 'Buy-Tokens-Error',
+            Error = 'Failed to mint tokens. Amount will be refunded.'
+        })
+
+        return
+    end
+
+    ReserveBalance = utils.add(ReserveBalance, quantityReservesSent)
+    if bint(ReserveBalance) >= bint(FundingGoal) then
```

Comment:
```suggestion
    if bint.__le(bint(FundingGoal), bint(ReserveBalance)) then
```

I would use `bint.__le` for comparisons